### ISSUE TITLE
Not working after Chrome cloud sync, possible firewall issue?

### ISSUE THREAD

**Issue by j-lo** [micjamking/devtools-author#86]
Hey Mike, great extension man, really loving it. I got it to work from home last night on two separate Chrome profiles, but I got in to work this morning and on the same two profiles the Author Settings won't work/appear properly, the Elements/Console etc. panels wouldn't apply the theme, and the Author Settings panel is mostly blank. Attached a screenshot of what it looks like, you can't select a theme or interact with the drop-down, and you can change the font size but it reverts back once you restart the dev tools. The funny thing is, it still seems to remember my theme selections from the previous night, because when you hover over anything in the Elements panel the corresponding web page highlights it in the theme's colours.

![screen shot 2016-01-29 at 8 40 04 am](https://cloud.githubusercontent.com/assets/5381716/12661793/71570db0-c667-11e5-81f2-e5f58beb60d6.png)

I did notice that while my Chrome profiles at work had copied over the extension and dev tools settings the flags still had the Enable Developer Tools experiments set to Enable, but even after enabling it I had no luck. I completely wiped everything, reset all settings, emptied the cache and started again, reinstalled the extension but still no love. I just tried a fresh install on a third separate Chrome profile and the problem still occurred. Just spitballing, but could this be a firewall issue that might prevent the extension from functioning properly? Wouldn't have thought so, have other extensions running.

**Comment by micjamking** [comment]
Hi @j-lo; I haven't seen this issue, however it is quite possible it could be related to a firewall conflict since the theme styling and preview are dynamically loaded using XHR. I can't say 100% that is the issues, since the plugin doesn't require internet access to run and the CSS and JSON assets are technically being served from the same domain.

What is your environment versions (Chrome, OS)?
Can you confirm if the issue exists on the same machine when accessing a different network?

Any other details you can provide will be extremely helpful in resolving the issue.

**Comment by j-lo** [comment]
Hey @micjamking thanks for getting on to this so quickly, apologies for breaking the extension so soon after use :)
Yeah that's what I suspected, I thought it could work independently of a connection.
So to clarify, I'm working off of Chrome 48.0.2564.97, on an iMac running El Capitan 10.11.3. I do have a few other extensions running, not sure if it's possible that they would interfere (just Colorzilla, Form Filler, Adobe Edge Inspect).
I tried a couple of different networks, patched into a local wi-fi one, and even tried with no network at all, same issue still occurs unfortunately.

**Comment by micjamking** [comment]
Hi @j-lo 

I'm assuming you've already followed the [usage guidelines](https://github.com/micjamking/devtools-author#usage) from the README since you mentioned you were able to get it running on two profiles from home (ie. enabled Developer Tools experiments and checked Allow custom UI themes in DevTools settings).

Does the problem occur on other machines in the same network, or isolated to your machine?

Can you inspect DevTools by undocking it to a separate window and then opening a subsequent DevTools window (`cmd + opt + I`) and seeing if there are any console errors?

Since I'm unable to replicate the issue, and I can't see any errors, this is going to be a tough one to crack :unamused:

**Comment by micjamking** [comment]
I can't say for sure, since I'm unable to replicate, but I have a feeling the issue may be due to my google analytics script blocking for some reason or another, within your environment. 

I've added an `async` attribute to the analytics script tag, fingers crossed this resolves the issue. Has been pushed out in release v0.4.8 https://github.com/micjamking/devtools-author/commit/a1f076dac5ae9a82f2702c5c34e7e44029ee0615

Can you confirm if you are still experiencing this issue after updating? I just pushed it out to the Chrome Store, so it could take up to an hour for the update to show up. Please let me know.

**Comment by j-lo** [comment]
Hey @micjamking, 

Yeah, I followed the guidelines. I just tried it on a workmate's PC and managed to get it running successfully, running Chrome 47.0.2526.111 on Windows 7 Pro, so it seems to just be my machine. And yeah, did a double inspection of the dev tools, no errors in the console.
The only thing/difference I've managed to pick up is that for some reason Chrome on my work machine has generated an extra option in the settings for the dev tools, named 'Blackboxing'. It wasn't there previously, it just popped up this morning - so it lists 'General', then 'Workspace', then 'Blackboxing'. Not sure if it's somehow related but it's the only difference I can find so far.

It's definitely a weird issue, appreciate the help all the same ;)

**Comment by j-lo** [comment]
Ah great work re the analytics fix @micjamking, I'll give it another try later this afternoon, will keep you posted on how I go.
Thanks again for the support with this.

**Comment by micjamking** [comment]
@j-lo any luck?

**Comment by j-lo** [comment]
Hey @micjamking no not yet mate, the Chrome Store hadn't updated by the time I had to finish up on Friday, am just waiting till I get back on the work computer to test it again, will let you know as soon as I can, probably Wednesday.

**Comment by j-lo** [comment]
@micjamking, downloaded the latest version, still no luck. Have a slightly different view now (can see the Font-Family, though still can't select fields) but I'm guessing that's just some of the new updates:
![screen shot 2016-02-02 at 8 14 52 am](https://cloud.githubusercontent.com/assets/5381716/12733201/3c2e614e-c985-11e5-9a46-b3886c444bf1.png)

I had the same error triggered on my home PC now too, something tripped it, as best as I could tell it came from the LiveReload extension I was running, after a quick restart it went back to normal.

**Comment by DuckyDisciple** [comment]
I'm having this same problem. It was working when I was at home, but at work I see the same screen. I can check into our firewall settings to see if I can find if my computer is being blocked somewhere.

**Comment by micjamking** [comment]
@DuckyDisciple thank you! this has been a really tricky one to debug because I cannot replicate it and it's happening in really isolated situations. Please let me know if you discover anything new.

**Comment by DuckyDisciple** [comment]
@micjamking Unfortunately I don't see anything being blocked on the firewall. I opened the dev tools in another window and then opened another dev tools (like you mentioned above) and the only thing that stood out was a warning about XMLHttpRequest on the main thread being depricated.

If it helps at all, I am on a Chromebook, version 48.0.2564.92 beta

**Comment by DuckyDisciple** [comment]
I just looked through the source code a bit more to figure out what is going on. One possible error point is that it the function is initialized with window.ga as the 3rd parameter. When I look up window.ga in the console it is undefined. I was hoping I could put in a breakpoint and test this out, but I can't figure out how to use a breakpoint since I can't open the 2nd dev tools and refresh the extension's page.

If you have any ideas that you want me to try out, I'm willing to help.

**Comment by micjamking** [comment]
@DuckyDisciple can you try and run/build the plugin locally, removing the GA script from panel.html to see if it resolves the issue? I originally assumed it might have something to do with GA (see https://github.com/micjamking/devtools-author/issues/86#issuecomment-176529420).

**Comment by DuckyDisciple** [comment]
@micjamking I ran the unpacked plugin locally, both with the ga script and with it commented out. Both times it looked the same.

**Comment by micjamking** [comment]
@DuckyDisciple that's disappointing :unamused: ; I was almost sure that was the issue. The plugin kickstarts with an `init()` function on https://github.com/micjamking/devtools-author/blob/master/app/scripts/panel.js#L277, which itself contains an XHR to grab the theme.json file and load the themes in to the select menu on the settings page. 

The rest of the page is loaded as a callback function within the success condition of the XHR, so if the XHR fire or successfully load, then I imagine you'd get the exact same experience both you and @j-lo are experiencing. But neither you or @j-lo mentioned any errors in the console, which you should have had if the XHR returned an error.

I'll try to do more testing as soon as I can.

**Comment by DuckyDisciple** [comment]
I did a little bit of debugging. It seems like it is reading the JSON file just fine. The problem is with the storage. Nothing is being logged in the console inside the callback function for the `storage.get(...)`. I logged `storage` right before that call and it just said `[object Object]`.

Hope that helps. And again, I'm free to try any ideas you have.

**Comment by micjamking** [comment]
Hey @DuckyDisciple & @j-lo 

Can you try adding a `console.log(object)` on the following line(s) and let me know the output:

https://github.com/micjamking/devtools-author/blob/master/app/scripts/panel.js#L236
https://github.com/micjamking/devtools-author/blob/master/app/scripts/panel.js#L250

You should get an object with a single property, either `'devtools-theme'` or `'devtools-fontFamily'` and the property itself should have a value containing a string (either the name of the theme or the name of the font-family).

I think we're getting closer. I think it might be a serialization issue, since `[object Object]` is also the output you get when you try to retrieve an object from localStorage that hasn't been serialized/deserialized. However, DevTools Author only stores simply strings and numbers for the theme name, font-family and font-size, and according to the documentation _"Primitive values such as numbers will serialize as expected."_ [https://developer.chrome.com/apps/storage#type-StorageArea](https://developer.chrome.com/apps/storage#type-StorageArea)

There are known [issues with Chrome storage and Date object serialization](https://code.google.com/p/chromium/issues/detail?id=161319), but that's the closest thing to a clue I've found so far.

**Comment by j-lo** [comment]
Hey @DuckyDisciple and @micjamking, just quickly to wanted to say great work and thanks a bunch for the support on this, apologies I haven't been able to provide more help, it's a little out of my field of knowledge.

Thought I'd just let you know I was trying it again this morning and it seems to be working now. I assumed the extension just auto updates, though I checked `chrome://extensions/`, which says it's running 0.5.9, the Chrome Store says it's up to 0.5.11, just for reference if you need it.
Initially when it loaded it seemed like it was half-working, so I could see the theme name and colour/text settings, unlike the blank panel before, but I was unable to select/interact with any of the inputs. After closing it and re-opening it again it was fully functioning. Tested it on a bunch of pages, both local and live, didn't have any problems, was able to use/select the inputs and change themes and values without any trouble.

Without knowing what work you've both put into it or if you're still updating it, would it be too pre-emptive to say that it's now all fixed? (fingers crossed)

**Comment by micjamking** [comment]
Hey @j-lo, glad to hear you finally got it working!

I'm holding off for a response from @DuckyDisciple since he mentioned similar issues.

I've had 20+ bug fix releases since this issue appeared, but they have almost all been on the CSS side. The only code recent code changes which could impact this were the `async` attribute in v0.4.8 and adding custom font-family support in v0.5.0.

Not sure if you guys tried this, but some users have mentioned on twitter that restarting Chrome was required to see the changes/themes take affect as well.

**Comment by DuckyDisciple** [comment]
@micjamking Both of those log as undefined

**Comment by DuckyDisciple** [comment]
Also, this issue was happening at home as well for me, so it looks like it was the update, not the location that caused this. Likewise, my Chromebook has been rebooted a few times and the issue is still happening.

**Comment by micjamking** [comment]
@DuckyDisciple :unamused: okay. My next thought is to see exactly _what_ is being saved in your Chrome storage by the extension.

Before the `init()` call at the end of the **panel.js** file, can you try and add the following:

**[app/scripts/panel.js](https://github.com/micjamking/devtools-author/blob/master/app/scripts/panel.js#L275)**
```
storage.get(null, function(object) {
    console.log(object);
});
```
This should output an object containing properties of everything being stored via DevTools Author in Chrome storage. Can you share the output of the above?

After running the above, can you try replacing the code with `storage.clear();`; this will remove anything that was stored via the extension the next time the Author Settings panel is accessed. After that, try re-adding the `storage.get(...)` code above and see if the output changes when firing up the app after clearing the storage.

**Comment by DuckyDisciple** [comment]
@micjamking Alright, now I'm confused. It's working but it isn't. I put that line in and it logged an object. The Author Settings tab is populated now. I put those other 2 logs in and they showed the correct values. So it looks like everything is good now....except the theme isn't being applied. I went to make sure the settings were all correct, but I don't even have the experiments tab in the devtools settings anymore. I'll try rebooting and see if that fixes anything.

**Comment by DuckyDisciple** [comment]
Yeah, after a reboot it came up and is working fine. I'm glad it's working, but I wish I could have figured out what the problem actually was. The weird thing is that I'm running off of the locally unpacked extension, so I wouldn't have got any updates you made, and I hadn't really rebooted Chrome since the last time I tried. I have no clue what changed.

**Comment by micjamking** [comment]
@DuckyDisciple can you confirm if the chrome store extension is working for you? Much like you, I'd like to know the real cause/problem, but if it's working for both of you guys without hiccup, we can chalk it up to Chrome sync issues.

**Comment by DuckyDisciple** [comment]
@micjamking It is working in the chrome store. After seeing it working on my unpacked extension, I removed that, added it from the store, and tested again and it was working.

**Comment by micjamking** [comment]
@j-lo & @DuckyDisciple okay, thank you guys for all your patience with trying to track down the issue. Since it sounds like it's been resolved for everyone, I'm going to go ahead and close this issue. However if you have any further issues feel free to leave a comment and I'll reopen this thread.

Mahalo!

**Comment by DuckyDisciple** [comment]
@micjamking Looks like the issue is back. When I saw that the theme was missing, I went back to the theme selector menu and it was grayed out like before. I just uninstalled the extension and loaded the unpacked extension that I was testing with before (version 0.5.6) and it is working fine. Not sure if you have heard of others having this problem. I can do some testing if you would like.

**Comment by j-lo** [comment]
Yeah, @micjamking and @DuckyDisciple, can also confirm I'm getting the issue again too. It's a little random though, I can get some pages to work, then they don't. I can close them and refresh and they'll work again, but sporadically. I notice too that if I open two tabs from the same local page I'm working from the first dev tools will work, but not the second or subsequent ones. As of now though I can't get it to work at all.

**Comment by micjamking** [comment]
Hey guys, I'm really bummed about this still being an issue. :disappointed: 

@DuckyDisciple can you try what I recommended in https://github.com/micjamking/devtools-author/issues/86#issuecomment-180132678, specifically the `storage.clear()` part, and let me know how that goes.

So I can confirm, what are the device/OS you both are having issues on? I know it may have been mentioned before, but we've gone back and forth a few times and I've lost track a bit. (cc: @j-lo)

Appreciate the feedback, hopefully we can find the actual culprit soon. 

:v:

**Comment by j-lo** [comment]
Hey @micjamking, all good, I feel you, it's a really strange (and annoying) issue. If it's any consolation this is the third plugin I've broken in recent weeks, must just be me, so don't feel too bad.

Just to recap, I'm getting the issue occurring on both home and work setups, so a MacPro (late 2012) and an iMac (late 2013), both running latest versions of El Capitan (as of writing this anyway, 10.11.3). And it's really random, it's been working on and off for me this morning, got it working on one site on a local dev environment, and then not on others. I've been trying everything under the sun to test and pinpoint any particular triggers but nothing so far sorry.

**Comment by micjamking** [comment]
@j-lo Can you try disabling any/all plugins except for Devtools Author and accessing the plugin via an incognito window. You'll need to enable the plugin for incognito mode in chrome://extensions

Testing it this way, we can atleast isolate it from any extranneous environment variables. Thanks again for your help :v:

**Comment by j-lo** [comment]
@micjamking Ah, didn't think to try it in Incognito mode...
Okay, turned off all my plugins (eight in all not including yours), allowed _DevTools Author_ to operate in Incognito mode, and sure enough the extension worked. Tested it on about half a dozen pages or so, both live and local dev ones, all good.

So where to next, should I try testing each of the other extensions one by one to see when it breaks?

**Comment by micjamking** [comment]
@j-lo yup, but you may need to enable them for incognito in the same way you did DevTools Author in the extension settings panel.

If everything's working and you can't confirm the specific issue, then check try to replicate the process in normal mode, enabling each plugin one by one.

Let me know how it goes!

**Comment by j-lo** [comment]
@micjamking Will do, will keep you posted ;)

**Comment by j-lo** [comment]
Okay, @micjamking and @DuckyDisciple, ran through a gamut of tests with my other plugins/extensions in Chrome and I _may_ have found a constant trigger for the issue. Just for posterity, here's a list of the extensions I'm running:

- Four (default, I think) Google ones, so Google Docs, Google Docs Offline, Google Sheets and Google Slides - not sure if these appear for all users, maybe just because I've used them
- Live Reload
- Form Filler
- ColorZilla
- Adobe Edge Inspect CC
- and the DevTools Author, obviously

When I tested all of them in Incognito tabs (while the checkbox to allow them in incognito was ticked), the DevTools Author worked without a hitch. Tried it with all of them on, one at a time, and various other combinations (half on/half off, every second one etc.), all no problems.

When I tried them in normal mode I got a slightly different result.

- One at a time was fine.
- All of them enabled, it didn't work.
- Tried them in groups, mixed results, was 50/50.
- I found that having all of the Google ones enabled wasn't a problem, the extension ran fine.
- Having all Google ones turned on, plus one of the other extensions was fine.
- Having all Google ones turned on, plus two other extensions, it won't work.

So basically it came down to the third-party extensions, or _non-Google_ ones let's say. Having one turned on, regardless of which one, was fine. But the second I turned on two or multiple third-party extensions, no good. So for example having Live Reload enabled on it's own was fine. But Live Reload _plus_ Form Filler, or plus ColorZilla, no good. And you could switch any of those four third-party extensions around in any order and it was always the same result: one is fine, two+ is no good.

I guess I should clarify, since this isn't my area of expertise, that I'm not saying for sure that the other extensions are 100% the cause of the issue, but they're definitely triggering whatever the root issue is.

Hope that helps a bit fella's, see how you go with that info ;)

**Comment by DuckyDisciple** [comment]
Mine seems to be working now. I'm using an HP Chromebook 14 on version 48.0.2564.92 beta. It was really weird because when I went to test it today, half of my theme was applied (some colors and the font size were correct) but the theme dropdown was still empty like before. I closed all of my windows and then went into incognito mode to test and it was working. Then I opened a normal window and found it working there too. Maybe it just needed a reboot of Chrome. If it happens again, I can test that first.

**Comment by j-lo** [comment]
Yeah, similarly, mine's working fine now too @micjamking. So weird.
Tested it as I did before in both normal/incognito mode, all fine.

**Comment by micjamking** [comment]
Hi @j-lo & @DuckyDisciple!

I know you mentioned that it started working for you both again a few weeks ago, but I just wanted to follow up and see if you were continuing to have issues. I've done a bit of refactoring in [v0.6.4](https://github.com/micjamking/devtools-author/releases/tag/v0.6.4) and I'm hoping that it may have addressed the issue you guys are experiencing.

Can you confirm if the latest version DevTools Author (v0.6.4) is working for both?

Much Mahalos!
:v:

**Comment by DuckyDisciple** [comment]
@micjamking Everything has been working fine for me. I just upgraded to the newest version in the webstore (v0.6.4) and that is working too. Thanks!

**Comment by j-lo** [comment]
@micjamking yeah it's all good. It had been on and off for me the past week or so, but likewise I upgraded it to 0.6.4, had to give Chrome a restart but now it's working just fine ;)

**Comment by micjamking** [comment]
Hi @j-lo & @DuckyDisciple! Thanks for the heads up guys! I'm going to close this issue for now, but feel free to comment if the issue resurfaces and I will reopen the issue.

Mahalo!
:v:
